### ISSUE TITLE
add the ability pin the same CID multiple times

### ISSUE THREAD

**Issue by RubenKelevra** [ipfs/ipfs-cluster#1182]
**Current behavior**

Currently, a CID can only be pinned once to a cluster. If the same pin will be added a second time, all meta-informations of the first pin will be overwritten without notice.

This behavior isn't following the _principle of least surprise_ and isn't documented in the `--help` text of this command.

**Describe the feature you are proposing**

I propose that we allow pinning a CID multiple times to a cluster, as long as the name is unique. A `pin ls <CID>` will then just return all pins for this CID, while a `pin rm <CID>` requires an `--all` flag if there are more than one pin of a CID, otherwise the `pin rm` will fail. `pin rm <CID> --name 'abc'` will remove the CID pin with the name 'abc'.

**Prospective benefits**

`pin add` will always result in a new entry in the cluster, while avoiding accidentally overwrite of metadata of another pin.

**Real-world use case**

I pin files from multiple datasets, which might have the same CID. If I don't check for an already existing pin, I would overwrite pins from the other dataset - not pretty, but probably there can be done a workaround.

The real issue comes when one dataset removes a file, and thus I want to remove a pin from the cluster, but there's no way to remove just the pin for the one dataset, without affecting the second dataset.

The only workaround is to switch from single file pins to recursive directory pinning. But this creates a lot of unnecessary IO if just one or two files of the ten thousands of files have been changed.

**Comment by hsanjuan** [comment]
Actually the principle of least surprise is that when you re-add or re-pin something that is already in the cluster, this triggers pinning for that item in the places where it needs to be pinned (#888, #912), rather than adding a second item.

To support this in some way means major architectural and api changes (things are indexed by CID). It's also weird that two things with the same CID are not the same thing in Cluster.

Supporting this would actually be akin to supporting multiple independent pinsets (that may or not share some pins). While I see the use, I have to say that I don't see this getting worked on for long.

**Comment by RubenKelevra** [comment]
@hsanjuan well, I'm currently evaluating how difficult it would be to implement #1146.

There are different choices how to design this since ipfs-cluster is meant to support just objects, and no file structures.

The idea is to get as close to concurrent write access as possible, which leads to the issue that it's impossible to add/change an object without accidentally altering a different object on the cluster.

Say you have two nodes with write access and both want to add file X with different names, there's no way to know which one will win, right?

How do I determine that I'm running a command which has successfully altered the cluster-pinset?

What will happen to the secondary cluster pin? Will the metadata be overruled or will the add fail, since there was changes which were conflicting?

**Comment by RubenKelevra** [comment]
TL;DR:

I like to be able to write files to the cluster without disturbing existing metadata OR get an error back and an ID which is unique to the current state, to be able to reference for an "update" command, which does fail if the supplied ID is wrong.

Otherwise I don't see a way to implement concurrent NFS access to the cluster without a chance of dataloss :/

**Comment by hsanjuan** [comment]
It's not dataloss, it's overwrite. First one operation writes, and the the other. The only thing cluster can do is ensure consistency, at least in CRDT land.

**Comment by hsn10** [comment]
This Can be done by adding reference counter property to CID

**Comment by hsanjuan** [comment]
> This Can be done by adding reference counter property to CID

It gets more complex if pin options are not the same (i.e. different names, replication factors for the same CID).
